`ChestOpenEvent`, `ChestBlock` + `ChestBlockActor` implementation, fixes

Se7en-dev opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

### Additions
- Implemented `ChestOpenEvent`, (hooking `ChestBlock::use`), triggered when a player opens a chest. [CANCELABLE]
- Implemented the `ChestBlock` class
- Implemented the `ChestBlockActor` class, along with a few functions : `isLargeChest()`, `getContainer()`, `openBy` (see docstrings for more info)

### Changes/Fixes
- Moved the `PistonAction` enum from `blockevent.ts` -> `block.ts`
- Reformatted the `blockevent` file to match other event files (event classes first, then their implementation)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have tested my changes and confirmed that they are working


### Additional notes
It seems ESLint is throwing errors since the merge commit, but I am not sure why :/

karikera commented:
importing is cycling.

for example.
```ts
// a.ts
import './b';
```
```ts
// b.ts
import './a';
```

Se7en-dev commented:
Yes I've seen

It seems it is caused by the `ChestBlockActor.openBy()` function, that uses `Player` as a parameter. Importing the class produces the dependency cycles. What should I do, move it (and other related functions) to a new file, or remove it ?

mdisprgm commented:
> Yes I've seen
> 
> It seems it is caused by the `ChestBlockActor.openBy()` function, that uses `Player` as a parameter. Importing the class produces the dependency cycles. What should I do, move it (and other related functions) to a new file, remove it, or is there another way ?

you can import only as type :
```js
import type { Player } from "./player";
```

Se7en-dev commented:
Oh that's right ! Thank you 🙏

Se7en-dev commented:
Updated changes :

- Removed `ChestBlockActor.getContainer()`, as `BlockActor.getContainer` already exists. my bad.
- Implemented `ChestPairEvent`, (hooking `ChestBlockActor::pairWith()`, triggered when 2 chests are paired [CANCELABLE]
- Added missing `decay`s
- Implemented the `mPosition` field in `BlockActor`, along with a getter function

mdisprgm commented:
I think `ChestBlock`, `ChestBlockActor` are also should be decayed, right? @karikera

karikera commented:
Yes, it would be better to decay them.
Please decay everything except Actors or NetworkIdentifiers.

And I recommend using the getter function instead of the offset field. it's hard to maintain.
Please use `BlockActor::getPosition` instead of `@nativeField(BlockPos, 0x2C)`.

Se7en-dev commented:
Thanks for the `decay` usage clarification, I wasn't sure when to use them.

I did not use the getter function because it wasn't showing in IDA, bc of the stripped pdb. But it does exist 😆I'm going to fix this, my bad.

karikera commented on bdsx/bds/implements.ts:
```diff
@@ -839,8 +839,23 @@ BlockActor.prototype.load = function(tag) {
+    pk.pos.set(this.getPosition());
+    pk.data.construct(nbtData);
+    player.sendNetworkPacket(pk);
+    nbtData.destruct();
```
the return value of `allcoateAndSave()` needs to call `dispose()` not `destruct()`.
`destruct()` does not `free` the instance's own memory.
but `allocate` things use `malloc` for allocating own memory.

and don't use `construct` for the already constructed field or use `destruct` before using it.

karikera commented:
Please be careful to touch the code format. it's not your code.
I wish to keep the contributor's code format. don't want to change the format without reason.

Se7en-dev commented:
> Please be careful to touch the code format. it's not your code. I wish to keep the contributor's code format. don't want to change the format without reason.

Is it because of ce1b4f0055cb7a555253fde86b78d2091cb59c49 ? I reformatted it to better match the other event files that followed this structure (the event classes first, then the implementation), I think it is easier to navigate through the file this way. (but I still can reverse the changes if needed)

karikera commented on bdsx/event_impl/blockevent.ts:
```diff
@@ -75,18 +142,6 @@ function onBlockPlace(blockSource:BlockSource, block:Block, blockPos:BlockPos, f
 }
 const _onBlockPlace = procHacker.hooking("BlockSource::mayPlace", bool_t, null, BlockSource, Block, BlockPos, int32_t, Actor, bool_t)(onBlockPlace);
 
-export enum PistonAction {
```
Please keep backward compatibility.
If you just delete them. It can break other plugins.
